src/common/type/pack.h

```diff
@@ -0,0 +1,505 @@
+/***********************************************************************************************************************************
+Pack Type
+
+The pack type encodes binary data compactly while still maintaining structure and strict typing. The idea is based on Thrift,
+ProtocolBuffers, and Avro, compared here: https://medium.com/better-programming/use-binary-encoding-instead-of-json-dec745ec09b6.
+The pack type has been further optimized to balance between purely in-memory structures and those intended to be passed via a
+protocol or saved in a file.
+
+Integers are stored with base-128 varint encoding which is equivalent to network byte order, i.e., the endianness of the sending and
+receiving host don't matter.
+
+The overall idea is similar to JSON but IDs are used instead of names, typing is more granular, and the representation is far more
+compact. A pack can readily be converted to JSON but the reverse is not as precise due to loose typing in JSON. A pack is a stream
+format, i.e. it is intended to be read in order from beginning to end.
+
+Fields in a pack are identified by IDs. A field ID is stored as a delta from the previous ID, which is very efficient, but means
+that reading from the middle is generally not practical. The size of the gap between field IDs is important -- a gap of 1 never
+incurs extra cost, but depending on the field type larger gaps may require additional bytes to store the field ID delta.
+
+NULLs are not stored in a pack and are therefore not typed. A NULL is essentially just a gap in the field IDs. Fields that are
+frequently NULL are best stored at the end of an object. When using .defaultNull in write functions a NULL will be written (by
+making a gap in the IDs) if the value matches the default. When using .defaultNull in read functions the default will be returned
+when the field is NULL (i.e. missing). The standard default is the C default for that type (e.g. bool = false, int = 0) but can be
+changed with the .defaultValue parameter.
```

Comment:
What do you think about clarifying the last, especially with respect to the WriteStr:
```
The standard default is the C default for that type (e.g. bool = false, int = 0) but 
can be changed with the .defaultValue parameter. For example, 
pckWriteBoolP(write, false) will write a 0 with an ID into the pack, but 
pckWriteBoolP(write, false, .defaultNull = true) will not write to the pack,
it will simply skip the ID. However, pckWriteStrP(packWrite, NULL) is not valid,
instead .defaultNull=true must also be passed.
```

Reply:
That looks fine to me.

Reply:
commit 6e24fbba2aa1b27b04f2c9df41730472e263c6db